File: app/src/main/java/hcm/ditagis/com/cholon/qlsc/async/QueryServiceFeatureTableAsync.java
package hcm.ditagis.com.cholon.qlsc.async;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.AsyncTask;

import com.esri.arcgisruntime.concurrent.ListenableFuture;
import com.esri.arcgisruntime.data.Feature;
import com.esri.arcgisruntime.data.FeatureQueryResult;
import com.esri.arcgisruntime.data.QueryParameters;
import com.esri.arcgisruntime.data.ServiceFeatureTable;

import java.util.Iterator;
import java.util.concurrent.ExecutionException;

import hcm.ditagis.com.cholon.qlsc.entities.DApplication;


/**
 * Created by ThanLe on 4/16/2018.
 */

public class QueryServiceFeatureTableAsync extends AsyncTask<QueryParameters, Feature, Void> {
    @SuppressLint("StaticFieldLeak")
    private Activity mActivity;
    @SuppressLint("StaticFieldLeak")
    private AsyncResponse mDelegate;
    private DApplication mApplication;
    private ServiceFeatureTable mServiceFeatureTable;

    public interface AsyncResponse {
        void processFinish(Feature output);
    }

    public QueryServiceFeatureTableAsync(Activity activity,
                                         ServiceFeatureTable serviceFeatureTable, AsyncResponse delegate) {
        this.mActivity = activity;
        this.mApplication = (DApplication) activity.getApplication();
        this.mServiceFeatureTable = serviceFeatureTable;
        this.mDelegate = delegate;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
    }

    @Override
    protected Void doInBackground(QueryParameters... params) {
        try {
            if (params != null && params.length > 0) {


                ListenableFuture<FeatureQueryResult> featureQueryResultListenableFuture = mServiceFeatureTable.queryFeaturesAsync(params[0], ServiceFeatureTable.QueryFeatureFields.LOAD_ALL);
                featureQueryResultListenableFuture.addDoneListener(() -> {
                    try {
                        FeatureQueryResult result = featureQueryResultListenableFuture.get();
                        Iterator iterator = result.iterator();

                        if (iterator.hasNext()) {
                            Feature feature = (Feature) iterator.next();
                            publishProgress(feature);
                        } else {
                            publishProgress();
                        }

                    } catch (InterruptedException | ExecutionException e) {
                        e.printStackTrace();
                        publishProgress();
                    }
                });
            } else publishProgress();
        } catch (Exception e) {
            publishProgress();
        }
        return null;
    }

    @Override
    protected void onProgressUpdate(Feature... values) {
        if (values == null) {
            mDelegate.processFinish(null);
        } else if (values.length > 0) mDelegate.processFinish(values[0]);
        else mDelegate.processFinish(null);
    }


    @Override
    protected void onPostExecute(Void result) {


    }

}

File: app/src/main/java/hcm/ditagis/com/cholon/qlsc/async/QueryServiceFeatureTableAsync.kt
package hcm.ditagis.com.cholon.qlsc.async

import android.annotation.SuppressLint
import android.app.Activity
import android.os.AsyncTask
import com.esri.arcgisruntime.data.Feature
import com.esri.arcgisruntime.data.QueryParameters
import com.esri.arcgisruntime.data.ServiceFeatureTable
import hcm.ditagis.com.cholon.qlsc.entities.DApplication
import java.util.concurrent.ExecutionException


/**
 * Created by ThanLe on 4/16/2018.
 */

class QueryServiceFeatureTableAsync(@field:SuppressLint("StaticFieldLeak")
                                    private val mActivity: Activity, @field:SuppressLint("StaticFieldLeak")
                                    private val mDelegate: AsyncResponse) : AsyncTask<QueryParameters, Feature, Void>() {
    private val mApplication: DApplication = mActivity.application as DApplication
    private val mServiceFeatureTable: ServiceFeatureTable

    interface AsyncResponse {
        fun processFinish(output: Feature?)
    }

    init {
        this.mServiceFeatureTable = mApplication.dFeatureLayer!!.layer.featureTable as ServiceFeatureTable
    }

    override fun onPreExecute() {
        super.onPreExecute()
    }

    override fun doInBackground(vararg params: QueryParameters): Void? {
        try {
            if (params != null && params.size > 0) {


                val featureQueryResultListenableFuture = mServiceFeatureTable.queryFeaturesAsync(params[0], ServiceFeatureTable.QueryFeatureFields.LOAD_ALL)
                featureQueryResultListenableFuture.addDoneListener {
                    try {
                        val result = featureQueryResultListenableFuture.get()
                        val iterator = result.iterator()

                        if (iterator.hasNext()) {
                            val feature = iterator.next() as Feature
                            publishProgress(feature)
                        } else {
                            publishProgress()
                        }

                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                        publishProgress()
                    } catch (e: ExecutionException) {
                        e.printStackTrace()
                        publishProgress()
                    }
                }
            } else
                publishProgress()
        } catch (e: Exception) {
            publishProgress()
        }

        return null
    }

    override fun onProgressUpdate(vararg values: Feature) {
        if (values.isNotEmpty())
            mDelegate.processFinish(values[0])
        else
            mDelegate.processFinish(null)
    }




}